solr/solr-ref-guide/src/neural-search.adoc

```diff
@@ -0,0 +1,320 @@
+See https://doi.org/10.1016/j.is.2013.10.006[Approximate nearest neighbor algorithm based on navigable small world graphs [2014]] and https://arxiv.org/abs/1603.09320[this paper [2018]] for details.
++
+It has the same meaning as `M` from the later paper.
++
+Accepted values:
+Integer.
+
+`hnswBeamWidth`::
++
+[%autowidth,frame=none]
+|===
+|Optional |Default: 100
+|===
++
+(advanced) This parameter is specific for the `Lucene90HnswVectorsFormat` codec format:
++
+It is the number of nearest neighbor candidates to track while searching the graph for each newly inserted node.
++
+See https://doi.org/10.1016/j.is.2013.10.006[Approximate nearest neighbor algorithm based on navigable small world graphs [2014]] and https://arxiv.org/abs/1603.09320[this paper [2018]] for details.
++
+It has the same meaning as `efConstruction` from the later paper.
++
+Accepted values:
+Integer.
+
+DenseVectorField supports the attributes: `indexed`, `stored`.
+
+*N.B.* currently multivalue is not supported
+
+Here's how a `DenseVectorField` should be indexed:
+
+[.dynamic-tabs]
+--
+[example.tab-pane#json]
+====
+[.tab-label]*JSON*
+[source,json]
+----
+[{ "id": "1",
+"vector": [1.0, 2.5, 3.7, 4.1]
+},
+{ "id": "2",
+"vector": [1.5, 5.5, 6.7, 65.1]
+}
+]
+----
+====
+
+[example.tab-pane#xml]
+====
+[.tab-label]*XML*
+[source,xml]
+----
+<add>
+<doc>
+<field name="id">1</field>
+<field name="vector">1.0</field>
+<field name="vector">2.5</field>
+<field name="vector">3.7</field>
+<field name="vector">4.1</field>
+</doc>
+<doc>
+<field name="id">2</field>
+<field name="vector">1.5</field>
+<field name="vector">5.5</field>
+<field name="vector">6.7</field>
+<field name="vector">65.1</field>
+</doc>
+</add>
+----
+====
+
+[example.tab-pane#solrj]
+====
+[.tab-label]*SolrJ*
+[source,java,indent=0]
+----
+final SolrClient client = getSolrClient();
+
+final SolrInputDocument d1 = new SolrInputDocument();
+d1.setField("id", "1");
+d1.setField("vector", Arrays.asList(1.0f, 2.5f, 3.7f, 4.1f));
+
+
+final SolrInputDocument d2 = new SolrInputDocument();
+d2.setField("id", "2");
+d2.setField("vector", Arrays.asList(1.5f, 5.5f, 6.7f, 65.1f));
+
+client.add(Arrays.asList(d1, d2));
+----
+====
+--
+
+== Query Time
+This is the list of Apache Solr query approaches designed to support Neural Search:
+
+=== knn Query Parser
+The `knn` K-Nearest Neighbors query parser allows to find the k-nearest documents to the target vector according to indexed dense vectors in the given field.
+
+It takes the following parameters:
+
+`f`::
++
+[%autowidth,frame=none]
+|===
+|Mandatory
+|===
++
+The DenseVectorField to search in.
+
+`topK`::
++
+[%autowidth,frame=none]
+|===
+|Optional |Default: 10
+|===
++
+How many k-nearest results to return.
+
+Here's how to run a KNN search:
+
+[source,text]
+&q={!knn f=vector topK=10}[1.0, 2.0, 3.0, 4.0]
+
+The search results retrieved are the K-nearest to the vector in input `[1.0, 2.0, 3.0, 4.0]`, ranked by the similarityFunction configured at indexing time.
```

Comment:
I'm not super familiar with neural search, but does the `similarityFunction` need to be defined at index time? In my head the math to compare vectors happens at query time - so the similarityFunction should be definable at query time?

I saw the note above about there are limitations of what `similarityFunction` is used based on how the vector values are indexed (ie: unit length vectors), but it doesn't sound like checking the types of vectors is enforced at index time. If the check of type of vectors was done at index time, then I can see why `similarityFunction` should be defined at index time instead of query time.

Reply:
Hi Kevin,
K-Nearest Neighbor in Apache Solr is backed by the Apache Lucene Approximate Nearest Neighbor implementation based on Hierarchical Navigable Small World Graph (you have the reference to the papers, in this contribution as well and in Lucene source code).
It's approximate and it relies on building index time graphs that are specific to the similarity(vector distance) function decided.

The Apache Solr org.apache.solr.schema.DenseVectorField builds the Lucene org.apache.lucene.document.KnnVectorField that requires the SimilarityFunction at index time.
The similarity function is used by the org.apache.lucene.codecs.lucene90.Lucene90HnswVectorsFormat in the 
org.apache.lucene.codecs.lucene90.Lucene90HnswVectorsWriter#writeGraph method.
The information is present in the org.apache.lucene.index.FieldInfo org.apache.lucene.index.FieldInfo#getVectorSimilarityFunction, passed by org.apache.lucene.index.IndexingChain#initializeFieldInfo and fetched by org.apache.lucene.index.IndexingChain.FieldSchema.
In the FieldSchema the property is set by org/apache/lucene/index/IndexingChain.java:827 from the org.apache.lucene.index.IndexableFieldType implementation org.apache.lucene.document.FieldType#setVectorDimensionsAndSimilarityFunction.
And with that method we reconnect to the org.apache.lucene.document.KnnVectorField#createType.
Sorry for the long post, but I hope it helps in the comprehension.

Reply:
Thanks makes a ton of sense.